base/abstractarray.jl

```diff
@@ -2628,90 +2628,84 @@ julia> mapslices(sum, a, dims = [1,2])
 ```
 """
 function mapslices(f, A::AbstractArray; dims)
-    if isempty(dims)
-        return map(f,A)
-    end
-    if !isa(dims, AbstractVector)
-        dims = [dims...]
-    end
-
-    dimsA = [axes(A)...]
-    ndimsA = ndims(A)
-    alldims = [1:ndimsA;]
-
-    otherdims = setdiff(alldims, dims)
-
-    idx = Any[first(ind) for ind in axes(A)]
-    itershape   = tuple(dimsA[otherdims]...)
-    for d in dims
-        idx[d] = Slice(axes(A, d))
-    end
+    isempty(dims) && return map(f, A)
 
     # Apply the function to the first slice in order to determine the next steps
-    Aslice = A[idx...]
+    idx1 = ntuple(d -> d in dims ? (:) : firstindex(A,d), ndims(A))
+    Aslice = A[idx1...]
     r1 = f(Aslice)
-    # In some cases, we can re-use the first slice for a dramatic performance
-    # increase. The slice itself must be mutable and the result cannot contain
-    # any mutable containers. The following errs on the side of being overly
-    # strict (#18570 & #21123).
-    safe_for_reuse = isa(Aslice, StridedArray) &&
-                     (isa(r1, Number) || (isa(r1, AbstractArray) && eltype(r1) <: Number))
 
-    # determine result size and allocate
-    Rsize = copy(dimsA)
-    # TODO: maybe support removing dimensions
-    if !isa(r1, AbstractArray) || ndims(r1) == 0
+    if r1 isa AbstractArray && ndims(r1) > 0
+        # Could easily allow for ndims(res1) > length(dims), at least when size is 1,
+        # but for now we don't, to match old behaviour:
+        ndims(r1) > length(dims) && throw(DimensionMismatch(
```

Comment:
This makes the following an error, as it is on master, but deleting it would ignore trailing dimensions in `f(x)` without a problem:
```
mapslices(hcat, rand(4,5), dims=1)
```
It would also be easy to make this work (as if `rand(4,5,1)`); at present it's a DimensionMismatch while on master it's a BoundsError:
```
mapslices(x -> x * x', rand(4,5), dims=(1,3))
```

Reply:
Updated to explicitly allow the former (trivial trailing dimensions) and produce a helpful error on the latter (nontrivial extra dimensions).